components/data_prefetch/data_prefetch.tsx

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React from 'react';
+import PQueue from 'p-queue';
+import {Channel} from 'mattermost-redux/types/channels';
+
+import {Constants} from 'utils/constants';
+import {loadProfilesForSidebar} from 'actions/user_actions.jsx';
+
+const queue = new PQueue({concurrency: 2});
+
+type Props = {
+    currentChannelId: string;
+    prefetchQueueObj: Record<string, string[]>;
+    prefetchRequestStatus: Record<string, any>;
+    unreadChannels: Channel[];
+    actions: {
+        prefetchChannelPosts: (channelId: string, delay: number | undefined) => Promise<any>;
+    };
+}
+
+export default class DataPrefetch extends React.PureComponent<Props, {}> {
+    private prefetchTimeout?: number;
+
+    async componentDidUpdate(prevProps: Props) {
+        const {currentChannelId, prefetchQueueObj} = this.props;
+        if (!prevProps.currentChannelId && currentChannelId) {
+            queue.add(async () => this.prefetchPosts(currentChannelId));
+            await loadProfilesForSidebar();
+            this.prefetchData();
+        } else if (prevProps.prefetchQueueObj !== prefetchQueueObj) {
+            clearTimeout(this.prefetchTimeout);
+            queue.clear();
+            this.prefetchTimeout = window.setTimeout(() => {
+                this.prefetchData();
+            }, 0);
+        }
+    }
+
+    public componentWillUnmount(): void {
+        clearTimeout(this.prefetchTimeout);
+    }
+
+    public prefetchPosts = (channelId: string) => {
+        let delay;
+        const channel = this.props.unreadChannels.find((unreadChannel) => channelId === unreadChannel.id);
```

Comment:
This looks like it's trying to fetch posts for the current channel when this is called from `componentDidUpdate`. Is that correct? Why would we need to do that when the PostList should already be doing that.

Reply:
We wouldn't fetch for the current channel because if there is a prefetch request in progress then it will not be made again. The if loop below in component prevents itt

```
if(this.props.prefetchRequestStatus[priorityQueue[channelIndex]] && this.shouldLoadPriorityQueue(priorityOrder, numberOfPiorityRequests))
```

Reply:
I don't mean that we'll have two prefetches being made for the current channel. I mean that we'll be making two requests for the posts: one prefetch and the other from the PostList.

Reply:
Nope we won't be doing that because postList won't be making a new request if prefetch is in progress

Reply:
Oh, okay. I missed that then.